src/MPQEditor/MPQCircleSelector.ts

```diff
@@ -0,0 +1,319 @@
+    args: MPQSelectionCmdOpenArgs,
+    handler: MPQSelectionEvent | undefined
+  ) {
+    super(extensionUri, panel.webview);
+
+    // let parsedPath = path.parse(args.docPath);
+    // let fileBase = path.join(parsedPath.dir, parsedPath.name);
+
+    this._panel = panel;
+    this._ownerId = args.document.uri.toString();
+    this._document = args.document;
+    this._modelPath = args.modelPath;
+    this._mpqEventHandler = handler;
+    this._lastSelected = args.names;
+
+    // Listen for when the panel is disposed
+    // This happens when the user closes the panel or when the panel is closed programmatically
+    this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+    // Update the content based on view changes
+    this._panel.onDidChangeViewState(
+      () => {
+        if (this._panel.visible) {
+          // NOTE if we call this.update(), it'll reload the model which may take time.
+          // TODO call conditional this.update() when necessary.
+          // this.update();
+        }
+      },
+      null,
+      this._disposables
+    );
+
+    this.initGraphCtrl(this._modelPath, this);
+    this.setMode("selector");
+  }
+
+  public dispose() {
+    this.disposeGraphCtrl();
+
+    MPQSelectionPanel.panels.forEach((selPan, index) => {
+      if (
+        this._ownerId === selPan._ownerId &&
+        this._modelPath === selPan._modelPath
+      ) {
+        MPQSelectionPanel.panels.splice(index, 1);
+      }
+    });
+
+    // Clean up our resources
+    this._panel.dispose();
+
+    while (this._disposables.length) {
+      const x = this._disposables.pop();
+      if (x) {
+        x.dispose();
+      }
+    }
+  }
+
+  /**
+   * CircleGraphEvent interface implementations
+   */
+  public onViewMessage(message: any) {
+    switch (message.command) {
+      case MessageDefs.selection:
+        this._lastSelected = message.names;
+        // we need to update the document, but not save to file.
+        // pass to owner to handle this.
+        if (this._mpqEventHandler) {
+          this._mpqEventHandler.onSelection(message.names, this._document);
+        }
+        break;
+      case MessageDefs.finishload:
+        this.onForwardSelection(this._lastSelected);
+        break;
+    }
+  }
+
+  public onForwardSelection(selection: any) {
+    let selections: string[] = [];
+    let items = selection as Array<string>;
+    for (let i = 0; i < items.length; i++) {
+      if (items[i].length > 0) {
+        selections.push(items[i]);
+      }
+    }
+    this.setSelection(selections);
+  }
+
+  public setTitle(title: string) {
+    this._panel.title = title;
+  }
+
+  public onLoadVisqInfo(visqPath: string) {
+    const fileData = fs.readFileSync(visqPath, { encoding: "utf8", flag: "r" });
+    const visq = JSON.parse(fileData);
+
+    // save selection
+    this.setMode("visqselector");
+    this._webview.postMessage({ command: "changemode", mode: "visqselector" });
+    this.sendVisq(visq);
```

Comment:
Although you explained in above comment about `changemode` but I am sill not confident about changing mode in runtime.

1/ you can change the scenario not to act like this, for example, when creating the json file, UI can ask the visq.json file. if provided open the `visqselector` mode, if not, `selector` mode.

2/ if you still must go with runtime `changemode`, changing mode to `visqselector` done in three calls doesn't seem good.
this may give users of `CircleGraphCtrl` make mistakes.
I think add an API in `CircleGraphCtrl` with `changeMode()` with some arguments would be better.

Reply:
> 1/ you can change the scenario not to act like this, for example, when creating the json file, UI can ask the visq.json file. if provided open the `visqselector` mode, if not, `selector` mode.

I agree. It will be safer to reload the html with proper mode as section above.